clar2wasm/src/words/equal.rs

```diff
@@ -67,3 +95,631 @@ impl Word for IsEq {
+        }
+        // is-eq-int function can be reused to both int and uint types.
+        TypeSignature::IntType | TypeSignature::UIntType => {
+            if ty == nth_ty {
+                wasm_equal_int128(generator, builder, first_op, nth_op)
+            } else {
+                no_type_match()
+            }
+        }
+        // is-eq-bytes function can be used for types with (offset, length)
+        TypeSignature::SequenceType(SequenceSubtype::BufferType(_))
+        | TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::ASCII(_)))
+        | TypeSignature::PrincipalType
+        | TypeSignature::CallableType(CallableSubtype::Principal(_)) => {
+            if ty == nth_ty {
+                wasm_equal_bytes(generator, builder, first_op, nth_op)
+            } else {
+                no_type_match()
+            }
+        }
+        TypeSignature::OptionalType(some_ty) => match nth_ty {
+            TypeSignature::OptionalType(nth_some_ty) => {
+                wasm_equal_optional(generator, builder, first_op, nth_op, some_ty, nth_some_ty)
+            }
+            _ => no_type_match(),
+        },
+        TypeSignature::ResponseType(ok_err_ty) => match nth_ty {
+            TypeSignature::ResponseType(nth_okerr_ty) => wasm_equal_response(
+                generator,
+                builder,
+                first_op,
+                nth_op,
+                (&ok_err_ty.0, &ok_err_ty.1),
+                (&nth_okerr_ty.0, &nth_okerr_ty.1),
+            ),
+            _ => no_type_match(),
+        },
+        TypeSignature::TupleType(tuple_ty) => match nth_ty {
+            TypeSignature::TupleType(nth_tuple_ty) => {
+                wasm_equal_tuple(generator, builder, first_op, nth_op, tuple_ty, nth_tuple_ty)
+            }
+            _ => no_type_match(),
+        },
+        TypeSignature::SequenceType(SequenceSubtype::ListType(list_ty)) => match nth_ty {
+            TypeSignature::SequenceType(SequenceSubtype::ListType(nth_list_ty)) => wasm_equal_list(
+                generator,
+                builder,
+                first_op,
+                nth_op,
+                list_ty.get_list_item_type(),
+                nth_list_ty.get_list_item_type(),
+            ),
+            _ => no_type_match(),
+        },
+        _ => Err(GeneratorError::NotImplemented),
+    }
+}
+
+fn wasm_equal_int128(
+    generator: &mut WasmGenerator,
+    builder: &mut InstrSeqBuilder,
+    first_op: &[LocalId],
+    nth_op: &[LocalId],
+) -> Result<(), GeneratorError> {
+    // Get first operand from the local and put it onto stack.
+    for val in first_op {
+        builder.local_get(*val);
+    }
+
+    // Get second operand from the local and put it onto stack.
+    for val in nth_op {
+        builder.local_get(*val);
+    }
+
+    // Call the function with the operands on the stack.
+    let func = OnceCell::new();
+    builder.call(*func.get_or_init(|| generator.func_by_name("stdlib.is-eq-int")));
```

Comment:
What do we gain from using `OnceCell` here (and other similar spots)?

Reply:
That's complete premature optimization.

The result of using `func_by_name` is cached for the future calls of this function. It's a bit like a `static` in C++.
I thought it was a good place to add it since we will make this call several times in this algorithm.

We can remove it if it bother you.

Reply:
Oh cool. I didn't realize it was kept across calls to the function, I thought it was just good for the lifetime of that `func` variable. Should `func` be `static` to make that work?

How does that compare with `lazy_static!`?

Maybe we should do this for all of the stdlib functions?

Reply:
No actually it's bs, I just checked my statements, it doesn't work like this. I will remove that thing.

But if you want more info about it, the page from the crate where it came from was interesting: [once_cell](https://docs.rs/once_cell/1.18.0/once_cell/).
It has been added to the standard lib since.

They say in their faq to not use `lazy_static` anymore, but no real reasons why. I think their main point is that they don't use macros.

Reply:
It seems like it could work with a static variable. But, sounds good, let's come back to this later, and when we decide how to do it, do it to all of these cases at once. 👍

Reply:
removed in 31b3aa88085c23434746f50564b7d3585e09a361

Reply:
Still, it's weird. Normally this should be called only the first time you invoke the function, but I added a print statement in the initialization and it's printed every time I call the function.

Reply:
Did you try making `func` static?

Reply:
yep, with the func static, it does work and only call the function once. But it causes the tests to break. Apparently it was not a correct assumption from my side that those functions would always have the same id.

Reply:
That's surprising. Anyway, let's merge this in without this optimization and revisit later. Thanks for digging into it!